azurelinuxagent/ga/extension_telemetry.py

```diff
@@ -0,0 +1,402 @@
+
+    def _parse_event_file_and_capture_events(self, handler_name, event_file_path, captured_events_count,
+                                             dropped_events_with_error_count):
+        events_list = []
+        event_file_time = datetime.datetime.fromtimestamp(os.path.getmtime(event_file_path))
+
+        # Read event file and decode it properly
+        with open(event_file_path, "rb") as fd:
+            event_data = fd.read().decode("utf-8")
+
+        # Parse the string and get the list of events
+        events = json.loads(event_data)
+
+        # We allow multiple events in a file but there can be an instance where the file only has a single
+        # JSON event and not a list. Handling that condition too
+        if not isinstance(events, list):
+            events = [events]
+
+        for event in events:
+            try:
+                events_list.append(self._parse_telemetry_event(handler_name, event, event_file_time))
+                captured_events_count += 1
+            except InvalidExtensionEventError as e:
+            # These are the errors thrown if there's an error parsing the event. We want to report these back to the
+            # extension publishers so that they are aware of the issues.
+            # The error messages are all static messages, we will use this to create a dict and emit an event at the
+            # end of each run to notify if there were any errors parsing events for the extension
+                dropped_events_with_error_count[ustr(e)] += 1
+            except Exception as e:
+                logger.warn("Unable to parse and transmit event, error: {0}".format(e))
+
+            if captured_events_count >= self._MAX_NUMBER_OF_EVENTS_PER_EXTENSION_PER_PERIOD:
+                break
+
+        return events_list
+
+    def _parse_telemetry_event(self, handler_name, extension_unparsed_event, event_file_time):
+        """
+        Parse the Json event file and convert it to TelemetryEvent object with the required data.
+        :return: Complete TelemetryEvent with all required fields filled up properly. Raises if event breaches contract.
+        """
+
+        extension_event = self._parse_event_and_ensure_it_is_valid(extension_unparsed_event)
+
+        # Create a telemetry event, add all common parameters to the event
+        # and then overwrite all the common params with extension events params if same
+
+        event = TelemetryEvent(TELEMETRY_LOG_EVENT_ID, TELEMETRY_LOG_PROVIDER_ID)
+        event.file_type = "json"
+        self.add_common_params_to_extension_event(event, event_file_time)
+
+        replace_or_add_params = {
+            GuestAgentGenericLogsSchema.EventName: "{0}-{1}".format(handler_name, extension_event[
+                ExtensionEventSchema.Version.lower()]),
+            GuestAgentGenericLogsSchema.CapabilityUsed: extension_event[ExtensionEventSchema.EventLevel.lower()],
```

Comment:
This is a bit confusing; are `EventLevel` and 'CapabilityUsed` equivalent names, or is this an instance of using an unused column for something else?

Reply:
This is basically to make sure that the Extension Telemetry Schema matches with the Kusto Table scema. The Schema in Extension Telemetry is easier for the author to understand, but the Kusto table schema is pre-set and can't be changed. To circumvent this we just mapped the columns in both the tables. More details in [this doc](https://microsoft.sharepoint.com/:w:/r/teams/ComputeVM/_layouts/15/Doc.aspx?sourcedoc=%7BD38D00FE-4D1E-49B3-A8D6-3FF97C21F6BB%7D&file=ExtensionTelemetryPipeline.docx&action=default&mobileredirect=true)